runtime/src/accounts_db.rs

```diff
@@ -2259,28 +2289,279 @@ impl AccountsDb {
+                    } else {
+                        // RPC get_account() may have fetched an old root from the index that was
+                        // either:
+                        // 1) Cleaned up by clean_accounts(), so the accounts index has been updated
+                        // and the storage entries have been removed.
+                        // 2) Dropped by purge_slots() because the slot was on a minor fork, which
+                        // removes the slots' storage entries but doesn't purge from the accounts index
+                        // (account index cleanup is left to clean for stored slots).
+                        num_acceptable_failed_iterations += 1;
+                    }
+                }
+            }
+            if num_acceptable_failed_iterations >= 10 {
+                // The latest version of the account existed in the index, but could not be
+                // fetched from storage. This means a race occurred between this function and clean
+                // accounts/purge_slots
+                let message = format!(
+                    "do_load() failed to get key: {} from storage, latest attempt was for \
+                     slot: {}, storage_entry: {} is_root_fixed: {}",
+                    pubkey, slot, store_id, is_root_fixed,
+                );
+                datapoint_warn!("accounts_db-do_load_warn", ("warn", message, String));
+                return None;
+            }
+
+            // Because reading from the cache/storage failed, retry from the index read
+            let (new_slot, new_store_id, new_offset) =
+                self.read_index_for_accessor(ancestors, pubkey, max_root)?;
+            if new_slot == slot && new_store_id == store_id {
+                // Considering that we're failed to get acccessor above and further that
+                // the index still returned the same (slot, store_id) tuple, offset must be same
+                // too.
+                assert!(new_offset == offset);
+
+                // If the entry was missing from the cache, that means it must have been flushed,
+                // and the accounts index is always updated before cache flush, so store_id must
+                // not indicate being cached.
+                assert!(new_store_id != CACHE_VIRTUAL_STORAGE_ID);
+
+                // If this is not a cache entry, then this was a minor fork slot
+                // that had its storage entries cleaned up by purge_slots() but hasn't been
+                // cleaned yet. That means this must be rpc access and not replay/banking.
+                // But we can't guarantee `is_root_fixed' must be false, because
+                // the dangling bank at the tip of minor fork might not be started to freeze
+                // (!freeze_started =~> is_root_fixed)
+
+                // Everything being said, let's return None as it's an error condition after all.
+                return None;
```

Comment:
I adjust these assertion ordering and commentary and added a test for this: https://github.com/solana-labs/solana/pull/15454/files?diff=split&w=1#r606630524